consensus/src/dag/round_state.rs

```diff
@@ -0,0 +1,154 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::dag::types::NodeCertificate;
+use aptos_consensus_types::common::Round;
+use aptos_types::epoch_state::EpochState;
+use async_trait::async_trait;
+use std::{
+    cmp::Ordering,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+/// Interface to decide if we should move forward to a new round
+#[async_trait]
+pub trait RoundState: Send {
+    async fn check_for_new_round(
+        &mut self,
+        highest_strong_links_round: Round,
+        strong_links: Vec<NodeCertificate>,
+    );
+
+    fn current_round(&self) -> Round;
+
+    fn set_current_round(&mut self, round: Round);
+}
+
+/// Move as fast as 2f+1
+pub struct OptimisticResponsiveRound {
+    current_round: Round,
+    event_sender: tokio::sync::mpsc::Sender<Round>,
+}
+
+impl OptimisticResponsiveRound {
+    pub fn new(event_sender: tokio::sync::mpsc::Sender<Round>) -> Self {
+        Self {
+            current_round: 0,
+            event_sender,
+        }
+    }
+}
+
+#[async_trait]
+impl RoundState for OptimisticResponsiveRound {
+    async fn check_for_new_round(
+        &mut self,
+        highest_strong_links_round: Round,
+        _strong_links: Vec<NodeCertificate>,
+    ) {
+        let new_round = highest_strong_links_round + 1;
+        if new_round > self.current_round {
+            self.event_sender
+                .send(new_round)
+                .await
+                .expect("Not able to send new round event");
+        }
+    }
+
+    fn current_round(&self) -> Round {
+        self.current_round
+    }
+
+    fn set_current_round(&mut self, round: Round) {
+        assert!(self.current_round < round);
+        self.current_round = round;
+    }
+}
+
+/// More sophisticated strategy to move round forward given 2f+1 strong links
+/// Delay if backpressure is triggered. (TODO)
+/// Move as soon as 3f+1 is ready. (TODO: make it configurable)
+/// Move if minimal wait time is reached. (TODO: make it configurable)
+pub struct AdaptiveResponsiveRound {
+    current_round: Round,
+    epoch_state: Arc<EpochState>,
+    start_time: Instant,
```

Comment:
The plan is to create new one every round?

Reply:
yeah i realized it after waking up 🤦